src/utils/reqwest_resume/mod.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+// Copyright 2018-2020 Alec Mocatta
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! Wrapper that uses the `Range` HTTP header to resume get requests.
+//!
+//! Most of the code can be attributed to `Alec Mocatta` and comes from the crate
+//! <https://crates.io/crates/reqwest_resume/>
+//! Some modifications have been done to update the code regarding `tokio`,
+//! replace the `hyperx` dependency with `hyper` and add two unit tests.
+
+use bytes::Bytes;
+use futures::{ready, FutureExt, Stream, TryFutureExt};
+use hyper::header::{self, HeaderMap, HeaderValue};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::sleep;
+
+/// Extension to [`reqwest::Client`] that provides a method to convert it
+pub trait ClientExt {
+    /// Convert a [`reqwest::Client`] into a [`reqwest_resume::Client`](Client)
+    fn resumable(self) -> Client;
+}
+impl ClientExt for reqwest::Client {
+    fn resumable(self) -> Client {
+        Client(self)
+    }
+}
+
+/// A `Client` to make Requests with.
+///
+/// See [`reqwest::Client`].
+#[derive(Debug)]
+pub struct Client(reqwest::Client);
+impl Client {
+    /// Constructs a new `Client`.
+    ///
+    /// See [`reqwest::Client::new()`].
+    pub fn new() -> Self {
+        Self(reqwest::Client::new())
+    }
+    /// Convenience method to make a `GET` request to a URL.
+    ///
+    /// See [`reqwest::Client::get()`].
+    pub fn get(&self, url: reqwest::Url) -> RequestBuilder {
+        // <U: reqwest::IntoUrl>
+        RequestBuilder(self.0.clone(), reqwest::Method::GET, url)
+    }
+}
+
+/// A builder to construct the properties of a Request.
+///
+/// See [`reqwest::RequestBuilder`].
+#[derive(Debug)]
+pub struct RequestBuilder(reqwest::Client, reqwest::Method, reqwest::Url);
+impl RequestBuilder {
+    /// Constructs the Request and sends it the target URL, returning a Response.
+    ///
+    /// See [`reqwest::RequestBuilder::send()`].
+    pub fn send(&mut self) -> impl Future<Output = reqwest::Result<Response>> + Send {
```

Comment:
Why is this not `async fn` instead of `fn -> impl Future`?

Reply:
Fixed.